src/librustc_resolve/diagnostics.rs

```diff
@@ -397,6 +397,102 @@ impl Bar {
 ```
 "##,
 
+E0412: r##"
+An undeclared type name was used. Example of erroneous codes:
```

Comment:
May want to say something about generics, e.g. `fn foo(x: T) {}` needing to be `fn foo<T>(x: T) {}`.